File: pulse-kotlin/src/commonTest/kotlin/com/fluxchess/pulse/kotlin/engine/SquareTest.kt
/*
 * Copyright 2013-2024 Phokham Nonava
 *
 * Use of this source code is governed by the MIT license that can be
 * found in the LICENSE file.
 */

package com.fluxchess.pulse.kotlin.engine

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class SquareTest {
	@Test
	fun `squares array should be valid`() {
		for (rank in ranks.indices) {
			for (file in files.indices) {
				assertEquals(squares[rank * ranks.size + file], squareOf(file, rank))
				assertTrue(isValidSquare(squareOf(file, rank)))
			}
		}
	}

	@Test
	fun `When creating a square it should save and return file and rank correctly`() {
		ranks.forEach { rank ->
			files.forEach { file ->
				val square = squareOf(file, rank)
				assertEquals(file, fileOf(square))
				assertEquals(rank, rankOf(square))
			}
		}
	}
}

File: pulse-go/internal/pulse/engine/square_test.go
/*
 * Copyright 2013-2024 Phokham Nonava
 *
 * Use of this source code is governed by the MIT license that can be
 * found in the LICENSE file.
 */

package engine

import (
	"testing"
)

func TestSquare(t *testing.T) {
	t.Run("Squares array should be valid", func(t *testing.T) {
		for r := range Ranks {
			for f := range Files {
				index := r*len(Ranks) + f
				sq := SquareOf(f, r)
				if Squares[index] != sq {
					t.Errorf("wanted Squares at index %d to be %v, but got %v", index, sq, Squares[index])
				}
				if !IsValidSquare(sq) {
					t.Errorf("wanted %v to be valid", sq)
				}
			}
		}
	})
}

func TestSquare_SquareOf(t *testing.T) {
	t.Run("When creating a square it should save and return file and rank correctly", func(t *testing.T) {
		for _, r := range Ranks {
			for _, f := range Files {
				sq := SquareOf(f, r)
				squareFile := FileOf(sq)
				if squareFile != f {
					t.Errorf("wanted file of square to be %v, but got %v", f, squareFile)
				}
				squareRank := RankOf(sq)
				if squareRank != r {
					t.Errorf("wanted rank of square to be %v, but got %v", r, squareRank)
				}
			}
		}
	})
}
